File: ruby/test/wire/compute_workspaces_test.rb
# frozen_string_literal: true

require_relative "wiremock_test_case"

class ComputeWorkspacesWireTest < WireMockTestCase
  def setup
    super

    @client = Sikaru::Client.new(
      token: "<token>",
      base_url: WIREMOCK_BASE_URL
    )
  end

  def test_compute_workspaces_get_with_wiremock
    test_id = "compute_workspaces.get.0"

    @client.compute_workspaces.get(
      project_id: "project_id",
      attachment_id: "attachment_id",
      run_id: "run_id",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "compute_workspaces.get.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/v1/projects/project_id/compute-attachments/attachment_id/workspace-checkpoints/run_id",
      query_params: nil,
      expected: 1
    )
  end

  def test_compute_workspaces_commit_tree_with_wiremock
    test_id = "compute_workspaces.commit_tree.0"

    @client.compute_workspaces.commit_tree(
      project_id: "project_id",
      attachment_id: "attachment_id",
      run_id: "run_id",
      files: {
        key: {
          chunks: [{
            sha256: "sha256",
            size: 1
          }],
          mode: 1,
          sha256: "sha256",
          size: 1
        }
      },
      request_options: {
        additional_headers: {
          "X-Test-Id" => "compute_workspaces.commit_tree.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/projects/project_id/compute-attachments/attachment_id/workspace-checkpoints/run_id/tree",
      query_params: nil,
      expected: 1
    )
  end
end

File: rust/tests/compute_workspaces_test.rs
use sikaru::prelude::*;

mod wire_test_utils;

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_compute_workspaces_get_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = Sikaru::new(config).expect("Failed to build client");

    let result = client
        .compute_workspaces
        .get(
            &"project_id".to_string(),
            &"attachment_id".to_string(),
            &"run_id".to_string(),
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "GET",
        "/v1/projects/project_id/compute-attachments/attachment_id/workspace-checkpoints/run_id",
        None,
        1,
    )
    .await
    .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_compute_workspaces_commit_tree_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = Sikaru::new(config).expect("Failed to build client");

    let result = client
        .compute_workspaces
        .commit_tree(
            &"project_id".to_string(),
            &"attachment_id".to_string(),
            &"run_id".to_string(),
            &WorkspaceTreeInput {
                files: HashMap::from([(
                    "key".to_string(),
                    WorkspaceFile {
                        chunks: vec![WorkspaceChunk {
                            sha256: "sha256".to_string(),
                            size: 1,
                            ..Default::default()
                        }],
                        mode: 1,
                        sha256: "sha256".to_string(),
                        size: 1,
                        ..Default::default()
                    },
                )]),
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("POST", "/v1/projects/project_id/compute-attachments/attachment_id/workspace-checkpoints/run_id/tree", None, 1).await.unwrap();
}
